GPU/Common/FragmentShaderGenerator.cpp

```diff
@@ -1078,16 +1082,37 @@ bool GenerateFragmentShader(const FShaderID &id, char *buffer, const ShaderLangu
 		return false;
 	}
 
-	// Final color computed - apply color write mask.
-	// TODO: Maybe optimize to only do math on the affected channels?
-	// Or .. meh. That would require more shader bits. Though we could
-	// of course optimize for the common mask 0xF00000, though again, blue-to-alpha
-	// does a better job with that.
-	if (colorWriteMask) {
+	// Final color computed - apply logic ops and bitwise color write mask, through shader blending, if specified.
+	if (colorWriteMask || replaceLogicOp) {
 		WRITE(p, "  highp uint v32 = packUnorm4x8(%s);\n", compat.fragColor0);
 		WRITE(p, "  highp uint d32 = packUnorm4x8(destColor);\n");
-		// Note that the mask has been flipped to the PC way - 1 means write.
-		WRITE(p, "  v32 = (v32 & u_colorWriteMask) | (d32 & ~u_colorWriteMask);\n");
+
+		// v32 is both the "s" to the logical operation, and the value that we'll merge to the destination with masking later.
+		// d32 is the "d" to the logical operation.
+		// TODO: Do logical ops work on just RGB or also A on the PSP?
```

Comment:
The software renderer code answers this question well, and it has been tested:

https://github.com/hrydgard/ppsspp/blob/748eef05d0c64b7aa5be6714fa52fa84a5977f05/GPU/Software/DrawPixel.cpp#L329

This was an added pain in the logicop jit, keeping the alpha bits from changing.

-[Unknown]

Reply:
As I think about this, I guess this might actually explain some broken dual-source alpha issues.  Hardware logic ops are actually unsafe generally, I guess, because of the stencil situation.

-[Unknown]

Reply:
Ah, right, I should have checked software. Interesting. Yeah maybe we should then simply never use them, and always fall back to in-shader, like I will have to do for D3D11.